Allow WebClient to be serialized and recreated without logging in again

dlenski opened the pull request:
As described in https://github.com/pR0Ps/stravaweblib/issues/2#issuecomment-932356236 and previously used in 44565582f56c6bffe6c2554193fbedb7ef58768f, logging into the Strava web interface creates 3 cookies:

- `strava_remember_token` is a [JWT](https://en.wikipedia.org/wiki/JSON_Web_Token)
- `strava_remember_id` is necessary for the web interface to recognize the session, but it is redundant since it can be recreated from the unencrypted payload of  `strava_remember_token`
- `_strava4_session` is unnecessary, because the web interface will still recognize the session if it's deleted

We can our understanding of the Strava cookies, and the CSRF token, to make it possible to easily serialize and recreate the WebClient without repeating the login process or retaining the user's password:

```py
>> from stravaweblib.webclient import WebClient
>> client = WebClient(email='me@my.com', password='password', access_token='StravaAPIToken')
>> print(repr(client))
stravaweblib.webclient.WebClient(csrf={'authenticity_token': '<base64 string>'}, strava_remember_token='<base64 Java Web Token>', access_token='StravaAPIToken')
>> print(c.client_args)
{'csrf': {'authenticity_token': '<base64 string>'}, 'strava_remember_token': '<base64 Java Web Token>', 'access_token': 'StravaAPIToken'}
>> args = c.client_args
>> del client
>> client = WebClient(**args)
```

Among other things, caching the client_args and reusing them in this way can avoid rate-limiting/lockout, which will happen if you log into the Strava web interface too frequently,

pR0Ps commented:
Sorry for the delay, I do intend to merge this, I just have to find the time to give it a spin.

dlenski commented:
Have you had a chance to test this?

pR0Ps commented:
Thanks for the reminder (and the original PR!) - I had definitely forgotten about this. I've added a commit with some changes.

I like using the JWT to resume sessions, I just wasn't a huge fan of the `client_args` method. Instead, I've added a property for the current JWT and made providing it to the `WebClient` a first-class option. I also refactored the login flow a bit, added some docs, and fixed up a few JWT-related edge cases that I ran into while testing (bad padding, expired).

dlenski commented:
> I like using the JWT to resume sessions, I just wasn't a huge fan of the client_args method. Instead, I've added a property for the current JWT and made providing it to the WebClient a first-class option.

Thanks for cleaning it up! Agreed, with your revisions it's more intuitive to use. :100:

> I also refactored the login flow a bit, added some docs, and fixed up a few JWT-related edge cases that I ran into while testing (bad padding, expired).

I never ran into a token with incorrectly-padded b64 from Strava while testing. Huh. :thinking:

pR0Ps commented:
> I never ran into a token with incorrectly-padded b64 from Strava while testing.

The only variable-length data in the JWT payload is the email and user id. The account(s) you tested must've all ended up with a payload length that perfectly aligned to 4 bytes. Luckily my account broke immediately or I probably would've missed it too haha
